apollo-federation/src/connectors/json_selection/apply_to.rs

```diff
@@ -708,7 +718,50 @@ impl ApplyToInternal for WithRange<PathList> {
         input_shape: Shape,
         dollar_shape: Shape,
     ) -> Shape {
-        match self.as_ref() {
+        if input_shape.is_none() {
+            // If the previous path prefix evaluated to None, path evaluation
+            // must terminate because there is no JSON value to pass as the
+            // input_shape to the rest of the path, so the output shape of the
+            // whole path must be None. Any errors that might explain an
+            // unexpected None value should already have been reported as
+            // Shape::error_with_partial errors at a higher level.
+            return input_shape;
+        }
+
+        match input_shape.case() {
+            ShapeCase::One(shapes) => {
+                return Shape::one(
+                    shapes.iter().map(|shape| {
+                        self.compute_output_shape(context, shape.clone(), dollar_shape.clone())
+                    }),
+                    input_shape.locations.iter().cloned(),
+                );
+            }
+            ShapeCase::All(shapes) => {
+                return Shape::all(
+                    shapes.iter().map(|shape| {
+                        self.compute_output_shape(context, shape.clone(), dollar_shape.clone())
+                    }),
+                    input_shape.locations.iter().cloned(),
+                );
+            }
+            ShapeCase::Error(error) => {
+                return match error.partial.as_ref() {
+                    Some(partial) => Shape::error_with_partial(
+                        error.message.clone(),
+                        self.compute_output_shape(context, partial.clone(), dollar_shape),
+                        input_shape.locations.clone(),
```

Comment:
```suggestion
                        input_shape.locations.iter().cloned(),
```

I think this form (used above as well) is more efficient, since the `Shape` constructor is going to call `into_iter()` and recollect again. So this way we reallocate the `Vec` once instead of twice.

We probably want to adjust the underlying `Shape` at some point to let us more easily move the `Vec` instead of constantly cloning it.

Reply:
This will change (get better) with https://github.com/apollographql/shape-rs/pull/29, after which you'd do something like `input_shape.locations()` for a `Iterator<Item = &Location>` or `input_shape.locations().cloned()` for `Iterator<Item = Location>`. Both locations and names are stored behind the new `Shape::meta` field.

Reply:
https://github.com/apollographql/router/pull/7932/commits/3705ee2c9a2eea3b7b7d0e08a3c8b870f28acbae